lib/trino-filesystem-alluxio/src/main/java/io/trino/filesystem/alluxio/AlluxioTrinoInputStream.java

```diff
@@ -0,0 +1,138 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.filesystem.alluxio;
+
+import alluxio.client.file.FileInStream;
+import io.trino.filesystem.TrinoInputStream;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Mostly copy pasted from AlluxioHdfsInputStream
+ * https://github.com/Alluxio/alluxio/blob/903269f077f30d3c274adfbc53ef248be8ff7356/core/client/hdfs/src/main/java/alluxio/hadoop/AlluxioHdfsInputStream.java#L26
+ */
+class AlluxioTrinoInputStream
+        extends FileInStream
+{
+    private final TrinoInputStream mInput;
+
+    public AlluxioTrinoInputStream(TrinoInputStream input)
+    {
+        this.mInput = requireNonNull(input);
+    }
+
+    @Override
+    public int read(byte[] bytes)
+            throws IOException
+    {
+        return this.mInput.read(bytes);
+    }
+
+    @Override
+    public int read(byte[] bytes, int offset, int length)
+            throws IOException
+    {
+        return this.mInput.read(bytes, offset, length);
+    }
+
+    @Override
+    public int read()
+            throws IOException
+    {
+        return this.mInput.read();
+    }
+
+    @Override
+    public int read(ByteBuffer buf)
+            throws IOException
+    {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public long skip(long length)
+            throws IOException
+    {
+        return this.mInput.skip(length);
+    }
+
+    @Override
+    public int available()
+            throws IOException
+    {
+        return this.mInput.available();
+    }
+
+    @Override
+    public void close()
+            throws IOException
+    {
+        this.mInput.close();
+    }
+
+    @Override
+    public synchronized void mark(int limit)
+    {
+        this.mInput.mark(limit);
+    }
+
+    @Override
+    public synchronized void reset()
+            throws IOException
+    {
+        this.mInput.reset();
+    }
+
+    @Override
+    public boolean markSupported()
+    {
+        return this.mInput.markSupported();
+    }
+
+    @Override
+    public void seek(long position)
+            throws IOException
+    {
+        this.mInput.seek(position);
+    }
+
+    @Override
+    public long getPos()
+            throws IOException
+    {
+        return this.mInput.getPosition();
+    }
+
+    @Override
+    public long remaining()
+    {
+        throw new UnsupportedOperationException("Remaining is not supported");
+    }
+
+    @Override
+    public int positionedRead(long position, byte[] buffer, int offset, int length)
+            throws IOException
+    {
+        throw new UnsupportedOperationException();
```

Comment:
This is definitely something that needs to be fixed in order for the cache to be generally usable. Otherwise, I think we'll see massive over-read. Could we handle this by subclassing or forking `LocalCacheFileInStream`?

My initial opinion after reading through `LocalCacheFileInStream` is that the code could be improved and is not a great fit for how `TrinoFileSystem` works. Having our own implementation that uses the lower level cache interfaces would allow us to directly call the underlying file system implementations in the correct and intended way.

Reply:
Thanks a lot for the suggestion! I've added a custom implementation of `AlluxioInput` which no longer delegates to `LocalCacheFileInStream` and with "native" implementations of `readFully` and `readTail`. We now only rely on Alluxio's `CacheManager` and some classes for keeping track of metadata. This has simplified the PR.

There are some open questions, and I'm not sure the current implementation of AlluxioInput is optimal. I'd highly appreciate feedback on the below points before I continue:
* `AlluxioInput.readFromExternalAndPopulateCache` allocates a byte array padded to Alluxio's page size, reads data into this byte array from the delegate file system, and then copies a subarray to the target array. This feels suboptimal, but I don't see an easy way to improve on it. From what I can tell, Alluxio does not support partial pages.
* `AlluxioInput.readFully` assumes that there will always be cache hits at the beginning of a read, and cache misses at the end of the read. We might want to support cache misses at the beginning of a read, and cache hits at the end as well?
* `AlluxioInputStream` still delegates to `LocalCacheFileInStream`. We might want to implement this in terms of `CacheManager`, but it looks like it going to be a bit more involved.
* We should add some metrics to keep track of the cache hit rate. It's probably better to add our own metrics, than to rely on what Alluxio does internally.

I've put off updating some of the Alluxio tests, as they will be affected by the above changes.

Reply:
To partially answer the first question. Instead of overriding `readFully(long position, byte[] buffer, int bufferOffset, int bufferLength)` of `TrinoInput` we could override `Slice readFully(long position, int length)`, since it can control the buffer which is used. This would mean refactoring most uses of `readFully(long position, byte[] buffer, int bufferOffset, int bufferLength)` to use `Slice readFully(long position, int length)`. Making such a refactor looks relatively easy, as most callers just allocate a fixed size buffer before calling `readFully`.

Reply:
Long term, we'd like to get rid of `Slice`, so it's best to not introduce any new usages of it. We should actually remove the `Slice` version of `readFully` from `TrinoInput` since it isn't used.

For now, we can simply do the copy in `AlluxioInput`. If it's a performance issue, we can work with Alluxio to improve it. Most of the time, the file readers will need to decompress the data, which is far more expensive than a memory copy, so I don't think we need to worry about this immediately.

Reply:
Thank you for the advice. I'll go with the current approach.

The main remaining item is to adopt `AlluxioInputStream` to not use `LocalCacheFileInStream`. We may also want to test `AlluxioInput` more carefully, even though existing test suites pass with the current implementation.

Reply:
This is implemented in https://github.com/trinodb/trino/pull/18719/commits/435cfa4ffa7ef1087f1bd301d5fb0f2035178f3f